Group reflection context docs together and provide new topic on behavior context (#1360)
o3de/o3de.org: content/docs/user-guide/programming/components/reflection/behavior-context.md

---
title: Behavior Context
description: Use the behavior context to make runtime code accessible to scripting systems such as Script Canvas or Lua.
weight: 300
---

The behavior context makes runtime code accessible to O3DE scripting systems such as Script Canvas or Lua. It provides script bindings that invoke runtime C++ methods, read constants, write properties, and create and handle EBus events. You can have multiple behavior contexts that are specialized for different purposes, and you can unreflect the behavior contexts to implement reloading.

Use the behavior context to bind the following C++ constructs for scripting:

+ [Classes](#classes)
+ [Methods](#methods)
+ [Properties](#properties)
+ [Constants](#constants)
+ [Enums](#enums)

In addition, the behavior context supports the O3DE [EBus](/docs/user-guide/programming/ebus) and [AZ::Event](/docs/user-guide/programming/az-event) event systems:

+ [EBus](#ebus)
  + [Events](#ebus-events)

>   + [Event Handlers](#ebus-event-handlers)

Author: Good catch! I must have changed my mind on those header titles. Thanks for testing.

## Classes

Reflects a C++ class or struct. You can provide an optional name. If you do not provide a name, the class name from `AzTypeInfo` is used. That name must be unique for the scope. Because the system uses `AzRTTI` to build the class hierarchy, you can use RTTI if you want to reflect base class functionality.

> Classes that are bound to the behavior context become objects that can be instantiated in a script environment. To reflect a class, you must provide the type that is reflected as a template argument to the class function.

Author: Sounds good to me. Suggestion updated.

```cpp
if (AZ::BehaviorContext* behaviorContext = azrtti_cast<AZ::BehaviorContext*>(context))
{
    behaviorcontext->Class<MyClass>();
}
```

The base class should also be specified, if applicable:

```cpp
if (AZ::BehaviorContext* behaviorContext = azrtti_cast<AZ::BehaviorContext*>(context))
{
    behaviorcontext->Class<MyClass, TheBaseClass>();
}
```

> ### Additional class information

Reviewer: Provide a link to the API reference page which describes how to set these properties, or provide a brief example somewhere in this section about how to do the relevant overrides.

Author: Good idea. Will do...

Author: Added.

> ### Additional class information

Author: ```suggestion
### Class reflection functions
```

+ **Allocator** - You can provide a custom allocator/ deallocator for your class. This allows you to override any existing allocation schema. If you do not provide a custom allocator, `aznew` is used (`AZ_CLASS_ALLOCATOR`).
+ **Constructor** - Allows you to enumerate the class constructors that you want to reflect. You must pass all constructor arguments as template arguments.
+ **Wrapping/WrappingMember** - Allows code to inform the system that it is a wrapper of another class. This is useful when you reflect smart pointers and string wrappers.
+ **Userdata** - Allows you to provide a pointer to user data. The pointer is accessible from all callbacks (like a custom allocator) that you implement for the class.
+ **Method** - Reflects a C++ class function. The first argument is the class pointer. This is the same usage as global methods.
+ **Property** - Reflects class data. The first argument is the class pointer. This is the same usage as global properties.
+ **Enum** - Enums are read-only `int` properties.
+ **Constant** - Constants are read-only properties.

### Attributes

> You can use the following, optional built-in attributes to decorate a class.

Reviewer: Suggest including an example showing how to set a variety of attributes during the context registration - seems to be a common pattern used in O3DE, especially for the editor context.

Author: Definitely. Will do...

Author: Added.

| Attribute | Description | Type | Values |
| --- | --- | --- | --- |

> | **Category** | Used by the editor to categorize the object in a list. You can use the forward slash (`/`) separator to nest categories: <br> `Attribute(AZ::Script::Attributes::Category, "Gameplay/Triggers")` | `string` | |

Author: Agreed. A lot of the lists in this topic were copied from the LY version of the doc. My fault for not revising.

| **ExcludeFrom** | An optional flag that hides this object from editor lists, self-documentation, preview builds, or all of the above. This flag is primarily used for internal objects that are not intended to be accessible by script. | `AZ::Script::Attributes::ExcludeFlags` | `List`, <br> `Documentation`, <br> `Preview`, <br> `All` |

> | **Storage** | Specifies the owner of the memory storage for the reflected object. Possible values include the script system, runtime native code, and stored by value in the script system's virtual machine (VM). | `AZ::Script::Attributes::StorageType` | `ScriptOwn`, <br> `RuntimeOwn`, <br> `Value` |

Author: Whew! Agreed. I'll examine to determine the original author's intent and then revise.

| **ClassNameOverride** | Provides a custom name for script reflection that is different from the behavior context name. | `string` | |
| **Ignore** | Specifies whether the element is ignored during reflection. | `bool` | `true`, <br> `false` |
| **Deprecated** | Marks a reflected class, method, EBus, or property as deprecated. | `bool` | `true`, <br> `false` |
| **ToolTip** | Used by the editor to display additional information in a tooltip. | `string` | |

### Lua usage examples

Classes reflected for scripting become objects that can be instanced in Lua:

```lua
local myObj = MyClass()
```

To use non-default constructors in Lua, first reflect the constructor to the behavior context:

```cpp
behaviorContext->Class<MyClass>("MyClass")
    ->Constructor<int>();
```

Then you can instantiate in Lua:

```lua
local myClass = MyClass(10)
```

### Using nested classes

To bind a nested class to the behavior context, you must do it from within a function on the nested class. C++ rules about nested class member access from outside scopes make this requirement necessary.

> The following counterexample shows a _poorly_ formed nested class. The code does not work.

Reviewer: If you're going to lead with the bad example, describe a bit more in the code itself via comments (or write some frontmatter/backmatter for the example) showing why it doesn't work, so that the correct code path can be emphasized later.

In general I'm wary about putting "don't do this" bad examples in the docs, but if this would be a common user mistake, no reason not to.

> The following counterexample shows a _poorly_ formed nested class. The code does not work.

Author: So...after taking a closer look at this "bad example" imported from the LY documentation, I couldn't say for sure what was actually "bad" about it. I decided to test it in a project and lo and behold, the "bad example" worked just fine! I ran it by an LY engineer and he, too, agreed there was nothing wrong with it. I don't know what the original contributor of this code was thinking, but I believe that I have proven it to be incorrect. I'm going to remove this section on nested classes.

```cpp
enum class MyTypes
{
    One = 1,
    Two = 2,
};

behaviorContext->Enum<aznumeric_cast<int>(MyTypes::One)>("MyTypes_One")
    ->Enum<aznumeric_cast<int>(MyTypes::Two)>("MyTypes_Two");
```

When enum values are reflected in this way, care should be taken to give each value a unique name, since each property must have a unique name in the behavior context.

To associate an enum with a class, invoke it on a `ClassBuilder` object that's returned by the `BehaviorContext::Class` function:

```cpp
behaviorContext->Class<PhotometricValue>("PhotometricUnit")
    ->Enum<static_cast<int>(PhotometricUnit::Candela>("Candela")
    ->Enum<static_cast<int>(PhotometricUnit::Lumen>("Lumen")
    ->Enum<static_cast<int>(PhotometricUnit::Lux>("Lux")
    ->Enum<static_cast<int>(PhotometricUnit::Unknown>("Unknown");
```

An associated class that is reflected with a name acts as a kind of namespace for the constant. For example, to get the value for `Lumen` from the preceding example in Lua, you would use `PhotometricUnit.Lumen`.

If a name is not specified for the associated class, no namespace prefix is needed.

### `EnumProperty` function

> The `EnumProperty` function is similar to `Enum`, helping you to define a getter for an enum in the behavior context. Additionally, because it returns a `GlobalPropertyBuilder` object, you can use it to associate additional attributes to the reflected enum:

Author: "associate...with" is accurate. Thanks!

```cpp
behaviorContext->EnumProperty<static_cast<int>(FrameCaptureResult::None)>("FrameCaptureResult_None")
    ->Attribute(AZ::Script::Attributes::Scope, AZ::Script::Attributes::ScopeFlags::Automation)
    ->Attribute(AZ::Script::Attributes::Module, "atom");
```

### `AZ_ENUM` utility macro

> O3DE `AzCore` provides a utility macro called `AZ_ENUM_DEFINE_REFLECT_UTILITIES` that works in conjunction with the `AZ_ENUM` macros to generate utility functions that can reflect _all_ values of an enum for you, saving you from the need to reflect each value individually. This might be useful for enums that are expected to change during the course of development, or that contain a large number of values. It requires the enum to be defined using either of the AZ_ENUM macros, including `AZ_ENUM`, `AZ_ENUM_WITH_UNDERLYING_TYPE`, `AZ_ENUM_CLASS`, or `AZ_ENUM_CLASS_WITH_UNDERLYING_TYPE`.

Author: Good point. I've updated your suggested change.

// Call the generated reflect function from within your component's Reflect function.
TestEnumReflect(*behaviorContext);
```

For additional context on how to use these macros, examine the test code in `Code\Framework\AzCore\Tests\EnumTests.cpp` or a Gem where they are used, such as `Gems\Atom\RHI\Code\Source\RHI.Reflect\RenderStates.cpp` and `Gems\Atom\RHI\Code\Include\Atom\RHI.Reflect\RenderStates.h`.

## EBus

Enable scripts to send and receive events by binding EBus events and event handlers to the behavior context:

### Events

EBus provides a mechanism to broadcast an event to all handlers or send an event directly to handlers connected at a specific ID. When you bind an event, O3DE automatically reflects `Broadcast`, `Event`, `QueueBroadcast`, and `QueueEvent`, as needed for your EBus configuration.

```cpp
behaviorContext.EBus<TestBus>("TestBus")->
    Handler<TestBusHandler>()->
    Event("SetSum1", &TestBus::Events::SetSum1)->
    Event("GetSum1", &TestBus::Events::GetSum1)->
;
```

Once compiled, the events can be used in Script Canvas and called in Lua:

```lua
local result = TestBus.Broadcast.GetSum1(1)
```

### Event handlers

> An event handler reflects a class that you must implement to forward messages from the EBus to behavior context methods. You must create a class that can monitor the specified EBus and forward messages to the behavior context. This is a requirement because the behavior context cannot guarantee that there is a handler for each message. If a message expects a result, you must provide a default result in case the message is not handled by the behavior context user. Keep in mind that the system creates as many of these handlers as the behavior context requires. Handlers can also execute in different threads. As a result, you should avoid static storage for values that change. The best way to understand this is to examine the example that follows.

Reviewer 2: ```suggestion
An event handler reflects a class that you must implement to forward messages from the EBus to behavior context methods. To implement the event handler, you must create a class that can monitor the specified EBus and forward messages to the behavior context.
 
This is required because the behavior context cannot guarantee that there is a handler for each message. If a message expects a result, you must provide a default result in case the behavior context user does not handle the message. Keep in mind that the system creates as many of these handlers as the behavior context requires. Handlers can also execute in different threads. As a result, you should avoid static storage for values that change. For example:
```
* For clarity, context, added "To implement the event handler..." in the second sentence. Verify for accuracy.
* Consider breaking up this long paragraph for easier scannability.
* Minor passive voice fix ("the message is not handled by the...user" to "the...user does not handle the message")
* I recommend avoid claiming that reviewing the following example is "The best way to understand this". You can probably just lead in with "For example:".

> 
> Example:

Reviewer 2: ```suggestion
```
I recommend cutting line 394 and a line above or below it, and going right into the following example.

Use the behavior context to bind the following C++ constructs for scripting:

+ [Classes](#classes)
+ [Methods](#methods)
+ [Properties](#properties)
+ [Constants](#constants)
+ [Enums](#enums)

In addition, the behavior context supports the O3DE [EBus](/docs/user-guide/programming/ebus) and [AZ::Event](/docs/user-guide/programming/az-event) event systems:

+ [EBus](#ebus)
  + [Events](#events)
  + [Event Handlers](#event-handlers)

## Classes

Classes in the behavior context reflect a C++ class or struct. You can provide an optional name for a class. If you do not provide a name, the class name from `AzTypeInfo` is used. That name must be unique for the scope. Because the system uses `AzRTTI` to build the class hierarchy, you can use RTTI if you want to reflect base class functionality.

Classes that you bind to the behavior context become objects that can be instantiated in a script environment. To reflect a class, you must provide the type that is reflected as a template argument to the class function. For example:

```cpp
if (AZ::BehaviorContext* behaviorContext = azrtti_cast<AZ::BehaviorContext*>(context))
{

>     behaviorcontext->Class<MyClass>();

Reviewer: ```suggestion
    behaviorContext->Class<MyClass>();
```

}
```

You should also specify the base class, if applicable:

```cpp
if (AZ::BehaviorContext* behaviorContext = azrtti_cast<AZ::BehaviorContext*>(context))
{

>     behaviorcontext->Class<MyClass, TheBaseClass>();

Reviewer: ```suggestion
    behaviorContext->Class<MyClass, TheBaseClass>();
```

* **Method** -- Reflects a C++ class function. The first argument is the class pointer. This is the same usage as global methods.
* **Property** -- Reflects class data. The first argument is the class pointer. This is the same usage as global properties.
* **Enum** -- Read-only `int` properties.
* **Constant** -- Read-only properties.

### Attributes

You can optionally use the following built-in attributes to decorate a class.

| Attribute | Description | Type | Values |
| --- | --- | --- | --- |
| **Category** | Used by the editor to categorize the object in a list. To nest categories, you can use the forward slash (`/`) separator. For example: <br> `Attribute(AZ::Script::Attributes::Category, "Gameplay/Triggers")` | `string` | |
| **ExcludeFrom** | An optional flag that hides this object from editor lists, self-documentation, preview builds, or all of the above. This flag is primarily used for internal objects that are not intended to be accessible by script. | `AZ::Script::Attributes::ExcludeFlags` | `List`, <br> `Documentation`, <br> `Preview`, <br> `All` |

> | **Storage** | Specifies the owner of the memory storage for the reflected object. Possible values include the script system, runtime native code, and stored by value in the script system's virtual machine (VM). | `AZ::Script::Attributes::StorageType` | `ScriptOwn`, <br> `RuntimeOwn`, <br> `Value` |

Author: ```suggestion
| **Storage** | Specifies the owner of the memory storage for the reflected object. An owner can be the script system (`ScriptOwn`), the native runtime code (`RuntimeOwn`), or the script system's virtual machine (`Value`). | `AZ::Script::Attributes::StorageType` | `ScriptOwn`, <br> `RuntimeOwn`, <br> `Value` |
```

| **ConstructibleFromNil** | Specifies whether the class is constructed by default when nil is provided. | `bool` | `true`, <br> `false` |
| **ClassNameOverride** | Provides a custom name for script reflection that is different from the behavior context name. | `string` | |
| **Ignore** | Specifies whether the element is ignored during reflection. | `bool` | `true`, <br> `false` |
| **Deprecated** | Marks a reflected class, method, EBus, or property as deprecated. | `bool` | `true`, <br> `false` |
| **ToolTip** | Used by the editor to display additional information in a tooltip. | `string` | |

### Lua usage examples

> Classes reflected for scripting become objects that can be instanced in Lua:

Author: ```suggestion
Classes reflected for scripting become objects that you can instance in Lua:
```

---
linkTitle: Behavior Context
title: Behavior Context in O3DE
description: Use the behavior context in Open 3D Engine (O3DE) to make runtime code accessible to scripting systems such as Script Canvas and Lua.
weight: 300
---

The behavior context makes runtime code accessible to **Open 3D Engine (O3DE)** scripting systems such as Script Canvas or Lua. It provides script bindings that invoke runtime C++ methods, read constants, write properties, and create and handle **Event Bus (Ebus)** events. You can have multiple behavior contexts that are specialized for different purposes, and you can unreflect the behavior contexts to implement reloading.

Use the behavior context to bind the following C++ constructs for scripting:

+ [Classes](#classes)
+ [Methods](#methods)
+ [Properties](#properties)
+ [Constants](#constants)
+ [Enums](#enums)

In addition, the behavior context supports the O3DE [EBus](/docs/user-guide/programming/ebus) and [AZ::Event](/docs/user-guide/programming/az-event) event systems:

+ [EBus](#ebus)
  + [Events](#events)
  + [Event Handlers](#event-handlers)

## Classes

> Classes in the behavior context reflect a C++ class or struct. You can provide an optional name for a class. If you do not provide a name, the class name from `AzTypeInfo` is used. That name must be unique for the scope. Because the system uses `AzRTTI` to build the class hierarchy, you can use RTTI if you want to reflect base class functionality.

Reviewer 2: ```suggestion
Classes in the behavior context reflect a C++ class or struct. You can provide an optional name for a class. If you don't provide a name, the class name from `AzTypeInfo` is used. That name must be unique for the scope. Because the system uses `AzRTTI` to build the class hierarchy, you can use RTTI if you want to reflect base class functionality.
```
Change for consistency with other occurrences of "don't" in the topic.

Classes that you bind to the behavior context become objects that can be instantiated in a script environment. To reflect a class, you must provide the type that is reflected as a template argument to the class function. For example:

```cpp
if (AZ::BehaviorContext* behaviorContext = azrtti_cast<AZ::BehaviorContext*>(context))
{
    behaviorContext->Class<MyClass>();
}
```

You should also specify the base class, if applicable:

```cpp
if (AZ::BehaviorContext* behaviorContext = azrtti_cast<AZ::BehaviorContext*>(context))
{
    behaviorContext->Class<MyClass, TheBaseClass>();
}
```

### Class builder functions

To provide additional configuration for the class binding, you can chain class builder functions after the behavior context `Class` function. The following builder functions are available:

| Builder Function | Description |
| --- | --- |

> | **Allocator** | Allows you to provide a custom allocator/deallocator for your class to override any existing allocation schema. If you do not provide a custom allocator, `aznew` is used (`AZ_CLASS_ALLOCATOR`). |

Reviewer 2: ```suggestion
| **Allocator** | Allows you to provide a custom allocator/deallocator for your class to override any existing allocation schema. If you don't provide a custom allocator, `aznew` is used (`AZ_CLASS_ALLOCATOR`). |
```
Change for consistency with other occurrences of "don't" in the topic.

> | **Constant** | Read-only properties. Refer to the section on [constants](#constants) for more information about this builder function. |
> | **Constructor** | Allows you to enumerate the class constructors that you want to reflect. You must pass all constructor arguments as template arguments. |
> | **Enum** | Read-only `int` properties. Refer to the section on [enums](#enums) for more information about this builder function. |
> | **Method** | Reflects a C++ class method. The function can also reflect global methods. Refer to the section on [methods](#methods) for more information about this builder function. |
> | **Property** | Reflects class data. The function can also reflect global properties. Refer to the section on [properties](#properties) for more information about this builder function. |
> | **UserData** | Allows you to provide a pointer to user data. This pointer is accessible from all callbacks (like a custom allocator) that you implement for the class. |

Reviewer 2: ```suggestion
| **Constant** | Read-only properties. For more information about this builder function, refer to the section on [constants](#constants). |
| **Constructor** | Allows you to enumerate the class constructors that you want to reflect. You must pass all constructor arguments as template arguments. |
| **Enum** | Read-only `int` properties. For more information about this builder function, refer to the section on [enums](#enums). |
| **Method** | Reflects a C++ class method. Can also reflect global methods. For more information about this builder function, refer to the section on [methods](#methods). |
| **Property** | Reflects class data. Can also reflect global properties. For more information about this builder function, refer to the section on [properties](#properties). |
| **UserData** | Allows you to provide a pointer to user data. This pointer is accessible from all callbacks (such as a custom allocator) that you implement for the class. |
```

#### C++ examples of class builder functions

```cpp
// Custom allocator and deallocator.
void* ScriptClassAllocate(void* userData);
void ScriptClassFree(void* obj, void* userData);

// Reflect the custom allocator and deallocator for use with ScriptClass.
behaviorContext->Class<ScriptClass>()
    ->Allocator(&ScriptClassAllocate, &ScriptClassFree);

// Reflect two constructors for BoxShapeConfig.
behaviorContext->Class<BoxShapeConfig>()
    ->Constructor()
    ->Constructor<AZ::Vector3&>()
    ->Property("Dimensions", BehaviorValueProperty(&BoxShapeConfig::m_dimensions));

// Allow BehaviorEntity to be passed to functions expecting AZ::Entity*.
behaviorContext->Class<BehaviorEntity>("Entity")
    ->WrappingMember<AZ::Entity*>(&BehaviorEntity::GetRawEntityPtr);

// Reflect a method, two properties, a constant, and an enum for ScriptClass.
behaviorContext->Class<ScriptClass>()
    ->Method("MemberFunc0", &ScriptClass::MemberFunc0)
    ->Property("data", &ScriptClass::GetData, &ScriptClass::SetData)
    ->Property("data1", BehaviorValueProperty(&ScriptClass::m_data1))
    ->Constant("EPSILON", BehaviorConstant(0.001f))
    ->Enum<ScriptClass::SC_ET_VALUE2>("SC_ET_VALUE2");
```

> Refer to [ClassBuilder](/docs/api/frameworks/azcore/struct_a_z_1_1_behavior_context_1_1_class_builder.html) in the O3DE API Reference for additional information about these functions.

Reviewer 2: Nice catch!

### Attributes

In addition to the builder functions, you can also use the following attributes to decorate a class.

| Attribute | Description | Type | Values |
| --- | --- | --- | --- |
| **Category** | Used by the editor to categorize the object in a list. To nest categories, you can use the forward slash (`/`) separator. For example: <br> `Attribute(AZ::Script::Attributes::Category, "Gameplay/Triggers")` | `string` | |
| **ClassNameOverride** | Provides a custom name for script reflection that is different from the behavior context name. | `string` | |

> | **ConstructibleFromNil** | Specifies whether the class is constructed by default when nil is provided. | `bool` | `true`, <br> `false` |
> | **ConstructorOverride** | Provide a custom constructor to be called when created from Lua script. | function pointer | |
> | **Deprecated** | Marks a reflected class, method, EBus, or property as deprecated. | `bool` | `true`, <br> `false` |
> | **ExcludeFrom** | An optional flag that hides this object from editor lists, self-documentation, preview builds, or all of the above. This flag is primarily used for internal objects that are not intended to be accessible by script. | `AZ::Script::Attributes::ExcludeFlags` | `List`, <br> `Documentation`, <br> `Preview`, <br> `All` |
> | **Ignore** | Specifies whether the element is ignored during reflection. | `bool` | `true`, <br> `false` |
> | **Storage** | Specifies the owner of the memory storage for the reflected object. An owner can be the script system (`ScriptOwn`), the native runtime code (`RuntimeOwn`), or the script system's virtual machine (`Value`). | `AZ::Script::Attributes::StorageType` | `ScriptOwn`, <br> `RuntimeOwn`, <br> `Value` |

Reviewer 2: ```suggestion
| **ConstructibleFromNil** | Specifies whether to construct the class by default when nil is provided. | `bool` | `true`, <br> `false` |
| **ConstructorOverride** | Provides a custom constructor to be called when created from Lua script. | function pointer | |
| **Deprecated** | Marks a reflected class, method, EBus, or property as deprecated. | `bool` | `true`, <br> `false` |
| **ExcludeFrom** | Hides the object from editor lists, self-documentation, preview builds, or all of the above. Use this optional flag primarily for internal objects that aren't intended to be accessible by script. | `AZ::Script::Attributes::ExcludeFlags` | `List`, <br> `Documentation`, <br> `Preview`, <br> `All` |
| **Ignore** | Specifies whether to ignore the element during reflection. | `bool` | `true`, <br> `false` |
| **Storage** | Specifies the owner of the memory storage for the reflected object. The owner can be the script system (`ScriptOwn`), the native runtime code (`RuntimeOwn`), or the script system's virtual machine (`Value`). | `AZ::Script::Attributes::StorageType` | `ScriptOwn`, <br> `RuntimeOwn`, <br> `Value` |
```
Misc. minor edits
